### ISSUE TITLE
Discuss: New Packaging Features?

### ISSUE THREAD

**Issue by ukadakal** [jxcore/jxcore#213]
Packaging is one of the most popular features in JXcore. What other functionalities can be included in packaging?

**Comment by ktrzeciaknubisa** [comment]
Personally I would like to have wildcard file/dir selection:

1. which files/dir to embed into a package

   During my jxcore testing adventure, so many times i wanted to pack just a single file which resides in directory together many other files, which I didn't want to have in a package. I could only copy the file into separate folder and pack from there.

    Speaking of this - there was a similar suggestion: #65 .

2. separate wildcards for `-slim` option
3. would be useful if all jxp fields could be defined from a command line, e.g.

    ````bash
$ jx package index.js Test -library -version "1.0" -author "Kris" -extract [and so on]
````

I think that's it if you ask me :)

**Comment by obastemur** [comment]
Apart from @ktrzeciaknubisa's suggestion (Kris can you handle that?) and stability updates, I don't think we can add more packaging features to current release. Maybe https://github.com/jxcore/jxcore/issues/181 but I don't know when I can prioritize that on my current load. Lets see.

**Comment by ktrzeciaknubisa** [comment]
@obastemur Sure. I'll do that asap and let you all guys know.

**Comment by ktrzeciaknubisa** [comment]
The first of three features mentioned above is implemented:

> 1. which files/dir to embed into a package

There is new `-add` option, which allows to specify files/folders to include into a package. For example the following command adds only *assets* folder (and *index.js* file naturally):

```bash
$ jx package index.js "Hello World" -add assets
```
It can still be combined with `-slim` option. Piece of documentation is [here](https://github.com/jxcore/jxcore/blob/master/doc/api/jxcore-feature-packaging-code-protection.markdown#-add). Relevant test case is also added to *test/jxcore/* folder.

**Comment by ktrzeciaknubisa** [comment]
The next of the features implemented:
> 2 . separate wildcards for -slim option

This is also applied for `-add` option.

The example below slims *node_modules* and adds (in addition to *index.js*) all txt files with name starting with "file" and located under *dir* folder:
```bash
$ jx package index.js "Hello World" -add "dir/file*.txt" -slim "node?modules"
```

Please note, that whenever you use wildcards, you should wrap given paths with a quotes, like in example above.

**Comment by ktrzeciaknubisa** [comment]
Last part of proposed features is also implemented:

> 3 . would be useful if all jxp fields could be defined from a command line

Example:

```bash
$ jx package index.js -library yes -version 1.0 -author "Kris" -extract no -preinstall "mkdir dir1"
```
In addition, the project name is no longer required. If not provided, will be taken from js file name. It will be just "index" in case of above example.

It's all documented [here](https://github.com/jxcore/jxcore/blob/master/doc/api/jxcore-feature-packaging-code-protection.markdown#package).

**Comment by ktrzeciaknubisa** [comment]
Another packaging feature mentioned in #370 is implemented : the partial contents extraction (44befcb7a344aac651a69174ff25fe522f07452c in 0.3.0.6 branch) supported by both command line params and *.jxp* files. 

New command line options for `jx package` are introduced (as extension to standard `-extract`, which btw. works the same): 
* `--extract-what`
* `--extract-where` 
* plus `--extract-app-root` (which is an alias to `--extract-where "./"`), .

For `--extract-what` wildcards and relative paths are supported (separated with comma or [JX_ARG_SEP](https://github.com/jxcore/jxcore/blob/0.3.0.6/doc/api/jxcore-utils.markdown#jxargsep)). 

Example:

```bash
$ jx package myapp.js --extract-what "*.txt,templates/dir" --extract-app-root
$ jx myapp.jx
```

Those params will be placed into *myapp.jxp* as such:

```js
"extract": {
		"what": [
			"*.txt",
			"templates/dir"
		],
		"where": "./"
}
```
This partial extraction feature is especially useful for running packages containing native modules (*.node* files), as we can mark them extractable.

Also extraction (full and partial) for native packages is enabled.

Documentation updated [here](https://github.com/jxcore/jxcore/blob/0.3.0.6/doc/api/jxcore-feature-packaging-code-protection.markdown#-extract).
